### PR TITLE
Implement isConnected method

### PR DESCRIPTION
**Pull request by bhartnett**
Having an isConnected method on the RpcClient will be useful for handling reconnection scenarios especially for websockets. At the moment if the websocket server goes offline then restarts the client is disconnected and not reconnected. Having an isConnected proc will allow checking and reconnecting in applications.

### CODE REVIEW

File: json_rpc/clients/httpclient.nim [hunk 1]
```diff
@@ -220,8 +220,12 @@ proc connect*(client: RpcHttpClient, address: string, port: Port, secure: bool)
   else:
     raise newException(RpcAddressUnresolvableError, res.error)
 
+method isConnected*(client: RpcHttpClient): bool =
+  (not client.httpSession.isNil()) and client.httpAddress.isOk()
+
 method close*(client: RpcHttpClient) {.async.} =
   if not client.httpSession.isNil:
     await client.httpSession.closeWait()
+    client.httpSession = nil
```
Comment (bhartnett) [inline json_rpc/clients/httpclient.nim:229]: It appears that this change has broken the CI for some reason. I'm looking into the cause.

Comment (jangko) [reply json_rpc/clients/httpclient.nim:229]: testhttps.nim too 
https://github.com/status-im/nim-json-rpc/blob/9da5b609bf27805ffaa598efd2db02c1eb5019f4/tests/testhttps.nim#L91

**Comment by jangko** [comment]
RpcHttpClient actually not using persistent connection like RpcWebsocketClient or RpcSocketClient. RpcHttpClient does not using keep-alive connection. Every request actually a new connection. I don't think we need to add isConnected to RpcHttpClient.

**Comment by bhartnett** [comment]
> RpcHttpClient actually not using persistent connection like RpcWebsocketClient or RpcSocketClient. RpcHttpClient does not using keep-alive connection. Every request actually a new connection. I don't think we need to add isConnected to RpcHttpClient.

I see, in that case it will be unimplemented and throw an assertion Defect if called on the RpcClient type. That might create a problem. For example, I've created a type that uses RpcClient and it might be a RpcHttpClient or a RpcWebsocketClient and I'll like to check the connection if an error is thrown in these cases with having to right code that checks what type of instance it is. 

Perhaps it would be better to simply return true for the RpcHttpClient? The RpcHttpClient does have a connect method already so perhaps isConnected in this context just means that it is ready to send requests.

File: tests/testserverclient.nim [hunk 2]
```diff
@@ -49,9 +49,51 @@ suite "Socket Server/Client RPC":
     except CatchableError as e:
       check e.msg == """{"code":-32001,"message":"Unknown payload"}"""
 
+  test "Client close and isConnected":
+    check client.isConnected() == true
+    # Is socket server close broken?
+    # waitFor client.close()
```
[the remark this answers is not in the document]
Comment (bhartnett) [reply tests/testserverclient.nim:55]: I didn't spend long looking at it. Do you have a suggestion/guidance on how I can fix it? If it isn't trivial, then I'll create an issue for it.

File: json_rpc/clients/httpclient.nim (hunk 1 above)
Comment (bhartnett) [reply json_rpc/clients/httpclient.nim:229]: I see, thanks. I'll fix.

**Comment by jangko** [comment]
> Perhaps it would be better to simply return true for the RpcHttpClient?

I have no objection, although it's a bit misleading technically. But I think it's ok. `isConnectedOrReadyToSend` is too wordy.

**Comment by bhartnett** [comment]
> > Perhaps it would be better to simply return true for the RpcHttpClient?
> 
> I have no objection, although it's a bit misleading technically. But I think it's ok. `isConnectedOrReadyToSend` is too wordy.

Yes thats true but perhaps if we implement keep alive then it won't matter. In that case, I'll leave it for now.

File: tests/testserverclient.nim (hunk 2 above)
[the remark this answers is not in the document]
Comment (bhartnett) [reply tests/testserverclient.nim]: I've just pushed a fix which solves the problem in the tests. Can you review the change when you get a chance? I'm not sure if it might effect something else.

**Review by jangko — APPROVED** [review]

**Comment by bhartnett** [comment]
> > Having an isConnected proc will allow checking and reconnecting in applications.
> 
> this is based on the assumption that connectivity accurately can be established, which is usually not true - ie the normal case is that you discover if you're connected or not on the first read or write from the socket, so the right place to handle reconnections is _usually_ in one of those two places: where you wait for new data to arrive and where you attempt to send stuff.

What would you suggest for this PR in that case? The overall goal for this PR was to have a way to decide whether to try reconnecting when using websockets and catching an error after issuing a json-rpc request. If the web socket connection is lost then a reconnect is required in order to continue sending data. 

In the RpcWebSocketClient type the WSSession transport field gets set to nil when the connection is lost and as you said this is only discovered when sending or receiving. Currently the only way for me to handle this is to check if the wsClient.transport field is nil which doesn't seem like a very nice solution.

**Comment by arnetheduck** [comment]
> What would you suggest for this PR in that case?

The "broad" way of handling this is to have the thing doing the requesting decide if it should retry based on the outcome of the request itself - the way sockets work is that many connection failures are detected when you try to use the socket, so this is usually the best place to "schedule" retries.

Wanting to check for `isConnected` _usually_ indicates that there is some underlying architectural issue, for example that it is not clear who "owns" the responsibility to open and close the connection when a request is being made.

**Comment by bhartnett** [comment]
> > What would you suggest for this PR in that case?
> 
> The "broad" way of handling this is to have the thing doing the requesting decide if it should retry based on the outcome of the request itself - the way sockets work is that many connection failures are detected when you try to use the socket, so this is usually the best place to "schedule" retries.
> 
> Wanting to check for `isConnected` _usually_ indicates that there is some underlying architectural issue, for example that it is not clear who "owns" the responsibility to open and close the connection when a request is being made.

It sounds like a better design for this might be to simply throw an exception at the time of doing a request which clearly indicates that a reconnect is required. The calling code can then catch and then handle appropriately. I'll close this PR for now and come up with a better solution.
